### ISSUE TITLE
Port twisted.internet.error to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5972]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5972|
|Type|enhancement|
|Created|2012-09-13 16:53:59Z|
|Branch|https://github.com/twisted/twisted/tree/ierror-py3-5972|

twisted.internet.error needs to run on Python 3.3.

<details><summary>Searchable metadata</summary>

```
trac-id__5972 5972
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_ierror_py3_5972 branches/ierror-py3-5972
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1347555239000000 1347555239000000
changetime__1347567528000000 1347567528000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [35629]) Branching to 'ierror-py3-5972'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst set owner to @exarkun|
|-|-|

The getConnectError API does not appear to be used well in Twisted, e.g. I'm pretty sure the `gaierror` code path is never hit. But, at least there's tests now. The changes to the code were because Python 3 doesn't have `__getitem__` on Exceptions anymore.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ierror-py3-5972 was started, haven't looked at it yet.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @itamarst<br>@exarkun set status to `new`|
|-|-|

  1. in `test_error.py`, please move the stdlib imports above the twisted imports.
  1. in `error.py`, perhaps the comment about Python 2.2 could be adjusted?  It seems like the author probably meant "Python 2.2 and newer", which effectively means all Python versions now.  I suspect `socket.gaierror` is always present now, too, but we can leave actually changing the logic for another time (file a ticket?)

Otherwise looks good, please merge after addressing the above.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

Will do. (I couldn't BTW find any instance where an actual exception was passed in, only tuples, but maybe I missed something.)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst set status to `closed`|
|-|-|

(In [35643]) Merge ierror-py3-5972

Author: itamar
Review: exarkun
Fixes: [#5972](https://github.com/twisted/twisted/issues/5972)

Port twisted.internet.error to Python 3, with improved test coverage.
